### PR TITLE
Improve expression resolving of superglobals

### PR DESCRIPTION
**Pull request by herndlm**
⚠️ This is a re-do of https://github.com/phpstan/phpstan-src/pull/2012 which caused performance issues. ⚠️

UPDATE: perf. testing is less relevant, the solution was adapted to lazily add global expression types which should avoid performance issues

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
I'm not a huge fan of this approach. It's going to pollute the output of `MutatingScope::debug()` (https://phpstan.org/r/e037be34-122b-4846-82f4-487dc0963845) and make it unreadable. Also the values are going to be in the scope and they are going to be mostly the same for all modern apps which isn't great.

Could we add these to the scope only when they are narrowed by some assignments or conditions? I think we already do a similar thing for `PHP_VERSION_ID` (maybe all constants). Thanks.

**Comment by herndlm** [comment]
> Could we add these to the scope only when they are narrowed by some assignments or conditions? I think we already do a similar thing for `PHP_VERSION_ID` (maybe all constants). Thanks.

smart, why haven't I thought about this :) I pushed something, not sure if it's going in the right direction or not though.

**Comment by herndlm** [comment]
I have a feeling that the stuff accessing `Scope::SUPERGLOBAL_VARIABLES` and `MutatingScope::isGlobalVariable()` as well as the looping in `getSuperglobalTypes()` and `getNativeSuperglobalTypes()` can be somehow improved still. I could try refactoring it.

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -5807,6 +5810,36 @@ public function getConstantReflection(Type $typeWithConstant, string $constantNa
...
+				continue;
+			}
+
+			$superglobalTypes[$exprString] = $typeHolder;
+		}
+		return $superglobalTypes;
+	}
+
+	/** @return array<string, ExpressionTypeHolder> */
+	private function getNativeSuperglobalTypes(): array
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:6009]: you could de-duplicate this method by passing in the expressionTypes via parameter

Comment (herndlm) [reply src/Analyser/MutatingScope.php:6009]: right. I stole this basically from `getNativeConstantTypes()` vs `getConstantTypes()` which could be adapted in a similar way then. maybe in a follow-up refactor :)

File: src/Analyser/MutatingScope.php [hunk 2]
```diff
@@ -3272,6 +3273,8 @@ public function enterNamespace(string $namespaceName): self
 			$this->isDeclareStrictTypes(),
 			null,
 			$namespaceName,
+			$this->getSuperglobalTypes(),
+			$this->getNativeSuperglobalTypes(),
 		);
```
Comment (herndlm) [inline src/Analyser/MutatingScope.php:3318]: I think this would not be needed anymore now

Comment (herndlm) [reply src/Analyser/MutatingScope.php:3318]: on the other hand.. removing this would be wrong

**Comment by herndlm** [comment]
https://github.com/phpstan/phpstan-src/pull/3871 is doing this in a better way. checked with the tests here which work there as well (except 2 edge cases I'm not sure and are most likely not worth to pursue)
